File: platform/backend/app/schemas/project.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ProjectType = Literal["spring-boot", "angular", "shared-lib", "mainframe", "python", "generic"]
PhaseType = Literal[
    "requirements", "architecture", "development", "testing", "cicd", "docs", "governance"
]
ProjectStatus = Literal["active", "paused", "archived"]


class ProjectCreate(BaseModel):
    """Payload for creating a new project."""

    model_config = ConfigDict(extra="forbid")

    organisation_id: uuid.UUID = Field(..., description="Owning organisation UUID")
    name: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(
        ...,
        min_length=1,
        max_length=255,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
    )
    description: str | None = Field(default=None)
    project_type: ProjectType = Field(default="generic")
    github_repo: str | None = Field(default=None, description="e.g. myorg/my-repo")
    jira_board_id: str | None = Field(default=None)
    confluence_space_key: str | None = Field(default=None)
    current_phase: PhaseType = Field(default="requirements")
    status: ProjectStatus = Field(default="active")


class ProjectUpdate(BaseModel):
    """Payload for partial-update of a project (all fields optional)."""

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=255)
    slug: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
        pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$",
    )
    description: str | None = Field(default=None)
    project_type: ProjectType | None = Field(default=None)
    github_repo: str | None = Field(default=None)
    jira_board_id: str | None = Field(default=None)
    confluence_space_key: str | None = Field(default=None)
    current_phase: PhaseType | None = Field(default=None)
    status: ProjectStatus | None = Field(default=None)


class ProjectResponse(BaseModel):
    """Read-side representation of a project."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organisation_id: uuid.UUID
    name: str
    slug: str
    description: str | None
    project_type: str
    github_repo: str | None
    jira_board_id: str | None
    confluence_space_key: str | None
    current_phase: str
    status: str
    created_at: datetime
    updated_at: datetime

File: platform/frontend/src/types/project.ts
import { z } from "zod";

export const ProjectTypeSchema = z.enum([
  "spring-boot",
  "angular",
  "shared-lib",
  "mainframe",
  "python",
  "generic",
]);
export type ProjectType = z.infer<typeof ProjectTypeSchema>;

export const PhaseTypeSchema = z.enum([
  "requirements",
  "architecture",
  "development",
  "testing",
  "cicd",
  "docs",
  "governance",
]);
export type PhaseType = z.infer<typeof PhaseTypeSchema>;

export const ProjectStatusSchema = z.enum(["active", "paused", "archived"]);
export type ProjectStatus = z.infer<typeof ProjectStatusSchema>;

export const ProjectSchema = z.object({
  id: z.string().uuid(),
  organisation_id: z.string().uuid(),
  name: z.string(),
  slug: z.string(),
  description: z.string().nullable(),
  project_type: ProjectTypeSchema,
  github_repo: z.string().nullable(),
  jira_board_id: z.string().nullable(),
  current_phase: PhaseTypeSchema,
  status: ProjectStatusSchema,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type Project = z.infer<typeof ProjectSchema>;

export const PersonaSchema = z.enum([
  "developer",
  "ba",
  "qa",
  "pm",
  "lead",
  "architect",
  "ciso",
]);
export type Persona = z.infer<typeof PersonaSchema>;

export const PHASE_ORDER: PhaseType[] = [
  "requirements",
  "architecture",
  "development",
  "testing",
  "cicd",
  "docs",
  "governance",
];

export const PHASE_LABELS: Record<PhaseType, string> = {
  requirements: "Requirements",
  architecture: "Architecture",
  development: "Development",
  testing: "Testing",
  cicd: "CI/CD",
  docs: "Docs",
  governance: "Governance",
};

export const PROJECT_TYPE_LABELS: Record<ProjectType, string> = {
  "spring-boot": "Spring Boot",
  angular: "Angular",
  "shared-lib": "Shared Lib",
  mainframe: "Mainframe",
  python: "Python",
  generic: "Generic",
};

export const PERSONA_LABELS: Record<Persona, string> = {
  developer: "Developer",
  ba: "Business Analyst",
  qa: "QA Engineer",
  pm: "Product Manager",
  lead: "Tech Lead",
  architect: "Architect",
  ciso: "CISO",
};
